DevUI pages for Hibernate ORM

TomasHofman opened the pull request:
* provide information about persistence units and their managed entities
* allow user to generate create-schema script

Fixes #16553

TomasHofman committed:
Hibernate ORM DevUI

* provide information about persistence units and their managed entities
* allow user to generate create-schema script

TomasHofman commented:
@Sanne I linked a PR with the first draft of the DevUI pages.

Vision: I think the useful things to show in the DevUI could be:

* a list of existing persistence units,
* a detail page for each PU showing a list of entities managed by given PU and a create-schema script.

Problems:

In order to generate the schema via the SchemaExport tool, I need to obtain the `Metadata` and `ServiceRegistry` instances of the persistence provider. I spend some time to figure out how to get them. I think my approach might not be entirely wrong but there are some further issues that need to be worked out.

* The implementation creates a new instance of `FastBootHibernatePersistenceProvider` to generate the schema and adds some non API methods to this class.
* The `PreconfiguredServiceRegistryBuilder#buildNewServiceRegistry()` method fails when called repeatedly (to generate new PersistenceProvider instance). I worked around it and added a comment to the place, but not sure if this is OK.
* The `PersistenceUnitsHolder#popRecordedState()` throws away the metadata after the PU are created, which makes it impossible to generate the `PersistenceProvider` repeatedly. Can the metadata be kept in dev-mode?

Sanne commented:
@TomasHofman  ! sorry for the delay, I've been off - need to catch up on things and then I'll have a better look at this. Thanks!

TomasHofman commented:
@Sanne  thanks and no problem, I'm catching up on other stuff too...

TomasHofman commented:
I did some more work on the HTML, attaching screenshots:

![embedded](https://user-images.githubusercontent.com/353221/116525826-32e26700-a8d9-11eb-8a22-0230f71776e2.png)
![persistence-units-ddls](https://user-images.githubusercontent.com/353221/116525829-34139400-a8d9-11eb-9e9b-9390236bee0a.png)
![entities](https://user-images.githubusercontent.com/353221/116525827-32e26700-a8d9-11eb-9b4a-71575f3e2c57.png)
![named-queries](https://user-images.githubusercontent.com/353221/116525828-337afd80-a8d9-11eb-9064-560b1b0f8e0d.png)

geoand commented:
I can't comment on any of this stuff, but having a DevUI entry for Hibernate ORM with such info  is an awesome idea!

Sanne commented on extensions/hibernate-orm/runtime/src/main/java/io/quarkus/hibernate/orm/runtime/PersistenceUnitsHolder.java:
```diff
@@ -54,7 +54,13 @@ public static RecordedState popRecordedState(String persistenceUnitName) {
+        //  The reason why the recorded state is removed here is to clear away the PU metadata after
+        //  the boot process, when they are no longer needed.
+        //  Would it be OK not to remove the metadata during the dev-mode execution?
+        return persistenceUnits.recordedStates.get(key);
```
The "remove" operation is actually important as the memory cost of keeping all metadata around is very large.

Do you think you could try keeping the state only when running in dev mode?

TomasHofman replied:
Yes, that's what I thought we could do. Yesterday I found out about `LaunchMode.current()` :), so I send another commit that would do `recordedStates.get(key)` or `recordedStates.remove(key)` here depending on current launch mode.

Sanne commented on extensions/hibernate-orm/runtime/src/main/java/io/quarkus/hibernate/orm/runtime/boot/registry/PreconfiguredServiceRegistryBuilder.java:
```diff
@@ -106,7 +106,13 @@ public StandardServiceRegistryImpl buildNewServiceRegistry() {
+        //  during persistenceProvider.generateSchema() execution (a new PersistenceProvider instance is constructed
+        //  during this call).
+        //  Is it OK to skip the resetAndReactivate() call when the registry is already active?
+        if (!destroyedRegistry.isActive()) {
```
this looks also a bit tricky, it should be possible to get ORM to generate the scripts without having to go through the bootstrap process of the registry.

I should probably try to find some time to help with this.

TomasHofman replied:
That would be great if you could look at that. I don't really have any insight into that code, I would most likely break it if I tried to refactor that bootstrap code :). Thanks Sanne!

yrodiere replied:
I pushed a commit to change how we retrieve the information: instead of being pulled at runtime, it's now generated on bootstrap. This works around all the problems related to generation of the schema after bootstrap.

Sanne reviewed (dismissed):
It's starting to look cool :)  But the ORM initialization code is a bit complex, I'll try to find some time to help with that.

I'm thinking there's a lot of functionality which we could add, but it's probably best to split it in smaller tasks.
For example, we could have:
 - also the "drop" script
 - a button to reset the DB state (actually runs the drop script, followed by the create and optionally the import.sql if there's any)

But IMO such additional things could be done via follow-up PRs, not least to get you some help from other people as well.
Let's focus first on the ORM boostrap changes and get the basics integrated first.

The "Names Query" table probably needs a column to show to which persistence unit it belongs; or alternatively maybe it's best to show multiple tables, one each for each PU.

TomasHofman commented:
I can squash the commits when we are done, but I will keep adding updates in separate commits for now to make it easier to review.

TomasHofman commented:
> The "Names Query" table probably needs a column to show to which persistence unit it belongs; or alternatively maybe it's best to 
> show multiple tables, one each for each PU.

Good idea, I would use the separate-table-for-each-PU approach for "Managed entities" page too, so that the pages are consistent.

yrodiere committed:
Hibernate ORM DevUI - Move dev-console-specific code to a dedicated package

yrodiere committed:
Hibernate ORM DevUI - Push info on startup instead of pulling it at runtime

yrodiere commented:
@TomasHofman @Sanne I pushed a commit to change how we retrieve the information: instead of being pulled at runtime, it's now generated on bootstrap. This works around all the problems related to generation of the schema after bootstrap.

Theoretically it should only have an impact in dev mode (not in tests nor in production), but let's wait for the build results :)

I also took the liberty of rebase this PR on the main branch since I don't think it makes sense to work on 1.13 right now, but let me know if that was a mistake.

@TomasHofman I know it's been a while, but do you think you could address the remaining comments? Feel free to ping me if you need help.

I believe y'all wanted to move to "one page per persistence unit", but after that's done I think it's going to be a great start and we can merge it?

TomasHofman commented:
@yrodiere that sounds great. I will take a look hopefully next week when I'm back from PTO and try to move it forward. Thanks!

Sanne commented:
Sounds great, many thanks @TomasHofman

TomasHofman committed:
Hibernate ORM DevUI - show separate table for each PU

TomasHofman commented:
Hello, I updated the PR with additional changes.

* Data for each PU are shown in separate tables now.
* Showing both create and drop SQL scripts for each PU.
* Some formatting changes.

I think this should cover all the requested changes, apart from providing a button to reset the database, which I would leave to a follow up PR.

Attaching current screenshots:

![overview](https://user-images.githubusercontent.com/353221/128671887-67c216b0-3a66-42b5-be5f-28be013d6697.png)
![pus](https://user-images.githubusercontent.com/353221/128671884-23770c62-63a8-404b-a8b9-74ba64fd219b.png)
![entities](https://user-images.githubusercontent.com/353221/128671880-b58ad6ac-14c6-4010-908d-2fc377c54e53.png)
![named-queries](https://user-images.githubusercontent.com/353221/128671882-05cd23d2-cc68-4630-8bb5-d766ef9ad044.png)

yrodiere committed:
Hibernate ORM DevUI - Make ORM => Vertx dependency optional

It's only there for DevUI

yrodiere committed:
Hibernate ORM DevUI - Don't keep the script selected after copying it.

It's annoying.

yrodiere reviewed (approved):
Thanks!

I pushed two more commits to address minor problems (one about dependencies, one about the javascript) and rebased it all on `main`.

Good to merge as soon as CI finishes the build.

Sanne commented:
Awesome, many thanks @TomasHofman and @yrodiere ! Lots of people will find this useful.

gsmet commented:
Would it be a problem to add a small test as done here: https://github.com/quarkusio/quarkus/pull/18738/files ?

The idea is mostly to check we don't break the pages with a 500 at some point.

TomasHofman commented:
I will have a look at the test.

TomasHofman committed:
Hibernate ORM DevUI - Smoke test

TomasHofman committed:
Hibernate ORM DevUI - Fix import file not being appended to DDL schema

The import file name was not correctly handed over to the SchemaExport
utility.

TomasHofman commented:
I added test test and a fix for a problem where the import.sql file name was not determined correctly. Waiting for CI to complete.

gsmet commented:
@TomasHofman small detail: could we change the ordering with a comparator to get the `<default>` one first and then the named ones in alphabetical order?

gsmet commented:
Also I wonder if I wouldn't make the scripts in the persistence units view foldable (and folded by default). It sorta works for a small db but will be more challenging with a bunch of entities. I **think** we have an example in the score page of the RESTEasy Reactive Dev UI.

I would add a simple icon to fold/unfold. It would work well with your existing UI, you could put the copy button and the folding button at the end of the `Create/Drop Script` line, that way, we could copy even if folded.

gsmet reviewed (approved):
Let's merge this, it's a very good step forward. We can tweak things in follow-up PRs. Thanks!

TomasHofman commented:
> Also I wonder if I wouldn't make the scripts in the persistence units view foldable (and folded by default). It sorta works for a small db but will be more challenging with a bunch of entities. I **think** we have an example in the score page of the RESTEasy Reactive Dev UI.
> 
> I would add a simple icon to fold/unfold. It would work well with your existing UI, you could put the copy button and the folding button at the end of the `Create/Drop Script` line, that way, we could copy even if folded.

Sounds like a good idea. I will create a new PR for that in couple of days after finish other work.

Thanks!
